### PR TITLE
Rec: Run Class/Interface checks on all entities

### PR DESCRIPTION
**Pull request by KnapSac**
Class and Interface checks now iterate over all entities, and can now
also contain multiple matches like a MethodCheck. This enables
recognizers to check for multiple entities and relations between the
entities.

To determine whether a ClassCheck isn't incorrectly nested, the parent
check is stored on the RecognizerContext. This context is refactored in
such a way that it is impossible to create a new instance outside of
RecognizerRunner.

As the initial root check doesn't have a CurrentEntity yet, and we don't
want to deal with that property being null everywhere, the RootNode
class is introduced which serves as a kind of sentinel value for the
root check.

### CODE REVIEW

File: PatternPal/PatternPal.Core/Checks/ICheck.cs [hunk 1]
```diff
@@ -115,22 +115,24 @@ internal static class CheckBuilder
     /// <param name="priority">The <see cref="Priority"/> of this <see cref="CheckCollection"/>.</param>
     /// <param name="checks">The sub-<see cref="ICheck"/>s of this <see cref="CheckCollection"/>.</param>
     /// <returns>The created <see cref="CheckCollection"/>.</returns>
-    internal static CheckCollection Any(
+    internal static NodeCheck< INode > Any(
```
Comment (CaptainJeroen) [inline PatternPal/PatternPal.Core/Checks/ICheck.cs:118]: This type is unhandled in the switchcase in RunCheck in NodeCheck.cs. This causes an exception thrown by the default case. You can recreate this error by running singletonTestcase11.cs through the consoleapp. Maybe add an extra case in the switch or there is a different type for any and all needed like the removed checkcollection?

**Review by CaptainJeroen — CHANGES_REQUESTED** [review]
Please add Any/All behaviour to the RunCheck in NodeCheck.cs, see comment in ICheck.cs

**Comment by KnapSac** [comment]
This behavior is implemented by the reduce results step, because while running the checks we don't know whether they're correct or not. That's why the kind is passed to the result.

File: PatternPal/PatternPal.Core/Checks/ICheck.cs (hunk 1 above)
Comment (KnapSac) [reply PatternPal/PatternPal.Core/Checks/ICheck.cs:118]: Good point, I'll add node check to the switch.

**Review by ViciousDoormat — APPROVED** [review]

**Pull request merged**